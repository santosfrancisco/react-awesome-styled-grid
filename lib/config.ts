import { Config, ConfigProps, GridBreakpoints, Media } from "@site/types";
import { css, CSSObject } from "styled-components";

export const CUSTOM_CONF = "awesomegrid";
export const DIMENSIONS: GridBreakpoints[] = ["xs", "sm", "md", "lg", "xl"];
export const BASE_CONF: Config = {
  media: null,
  mediaQuery: "only screen",
  columns: {
    xs: 4,
    sm: 8,
    md: 8,
    lg: 12,
    xl: 12
  },
  gutterWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  paddingWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  container: {
    xs: "full", // 'full' = max-width: 100%
    sm: "full", // 'full' = max-width: 100%
    md: "full", // 'full' = max-width: 100%
    lg: 90, // max-width: 1440px
    xl: 90 // max-width: 1440px
  },
  breakpoints: {
    xs: 1,
    sm: 48, // 768px
    md: 64, // 1024px
    lg: 90, // 1440px
    xl: 120 // 1920px
  }
};

const configs = [];
const hasCustomConf = (props: ConfigProps) => {
  if (!props?.theme) return {};
  return JSON.stringify(props.theme[CUSTOM_CONF] || {});
};
const resolveConfig = (props: ConfigProps) => {
  let themeConf = {};
  if (props?.theme) {
    themeConf = props.theme[CUSTOM_CONF] || {};
  }

  const conf: Config = {
    ...BASE_CONF,
    ...themeConf
  };

  conf.media = Object.keys(conf.breakpoints).reduce((media, breakpoint) => {
    const breakpointWidth = conf.breakpoints[breakpoint];
    media[breakpoint] = makeMedia(
      [
        conf.mediaQuery,
        breakpointWidth >= 0 && `(min-width: ${breakpointWidth}rem)`
      ]
        .filter(Boolean)
        .join(" and ")
    );
    return media;
  }, {} as Media);

  return conf;
};

export default function config(props?: ConfigProps): Config {
  const customConf = hasCustomConf(props);
  if (configs[0] === customConf) {
    return configs[1];
  }

  const conf = resolveConfig(props);

  configs[0] = customConf;
  configs[1] = conf;

  return conf;
}

function makeMedia(media: string) {
  return (...args: CSSObject[] | TemplateStringsArray) => css`
    @media ${media} {
      ${css.apply(null, [...args])}
    }
  `;
}
