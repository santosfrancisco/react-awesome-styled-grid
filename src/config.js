import { css } from 'styled-components'

const CUSTOM_CONF = 'awesomegrid'
export const DIMENSIONS = ['xs', 'sm', 'md', 'lg', 'xl']
export const BASE_CONF = {
  mediaQuery: 'only screen',
  columns: {
    xs: 4,
    sm: 8,
    md: 8,
    lg: 12,
    xl: 12
  },
  gutterWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  paddingWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  container: {
    xs: 'full', // 'full' = max-width: 100%
    sm: 'full', // 'full' = max-width: 100%
    md: 'full', // 'full' = max-width: 100%
    lg: 90, // max-width: 1440px
    xl: 90 // max-width: 1440px
  },
  breakpoints: {
    xs: 1,
    sm: 48, // 768px
    md: 64, // 1024px
    lg: 90, // 1440px
    xl: 120 // 1920px
  }
}

const configs = []
const hasCustomConf = props => JSON.stringify((props.theme && props.theme[CUSTOM_CONF]) || {})
const resolveConfig = props => {
  const themeConf = (props.theme && props.theme[CUSTOM_CONF]) || {}

  const conf = {
    ...BASE_CONF,
    ...themeConf
  }

  conf.media = Object.keys(conf.breakpoints).reduce((media, breakpoint) => {
    const breakpointWidth = conf.breakpoints[breakpoint]
    media[breakpoint] = makeMedia(
      [
        conf.mediaQuery,
        breakpointWidth >= 0 && `(min-width: ${breakpointWidth}rem)`
      ]
        .filter(Boolean)
        .join(' and ')
    )
    return media
  }, {})

  return conf
}

export default function config (props = {}) {
  const customConf = hasCustomConf(props)
  if (configs[0] === customConf) {
    return configs[1]
  }

  const conf = resolveConfig(props)

  configs[0] = customConf
  configs[1] = conf

  return conf
}

function makeMedia (media) {
  return (...args) => css`
    @media ${media} {
      ${css(...args)}
    }
  `
}
