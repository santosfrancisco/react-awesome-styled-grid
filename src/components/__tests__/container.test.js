import React from 'react'
import renderer from 'react-test-renderer'
import { DIMENSIONS, BASE_CONF } from '../../config'
import Container from '../grid/container'
import 'jest-styled-components'

describe('<Container />', () => {
  it('should render default style correctly', () => {
    const tree = renderer.create(<Container />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should haven`t a max-width when it is a fluid container', () => {
    const tree = renderer.create(<Container fluid />).toJSON()
    expect(tree).toHaveStyleRule('max-width', undefined)
  })

  it('should have a max-width when it is not a fluid container', () => {
    const tree = renderer.create(<Container />).toJSON()

    const { breakpoints, container } = BASE_CONF

    DIMENSIONS.forEach(d => {
      if (d === DIMENSIONS[0]) {
        expect(tree).toHaveStyleRule('max-width', undefined)
      } else {
        expect(tree).toHaveStyleRule(
          'max-width', `${container[d]}px`, {
            media: `only screen and (min-width: ${breakpoints[d]}px)`
          }
        )
      }
    })
  })

  it('should have different style when it`s debug props is true', () => {
    const tree = renderer.create(<Container debug />).toJSON()
    expect(tree).toHaveStyleRule('background-color', '#5901ad40')  
    expect(tree).toHaveStyleRule('outline', '#fff solid 1px')
  })
})