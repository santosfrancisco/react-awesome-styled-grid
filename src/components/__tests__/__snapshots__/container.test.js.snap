// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> should render default style correctly 1`] = `
.c0 {
  margin-right: auto;
  margin-left: auto;
}

@media only screen and (min-width:1rem) {
  .c0 {
    padding-left: 1rem;
    padding-right: 1rem;
  }
}

@media only screen and (min-width:48rem) {
  .c0 {
    padding-left: 1rem;
    padding-right: 1rem;
  }
}

@media only screen and (min-width:64rem) {
  .c0 {
    padding-left: 1.5rem;
    padding-right: 1.5rem;
  }
}

@media only screen and (min-width:90rem) {
  .c0 {
    padding-left: 1.5rem;
    padding-right: 1.5rem;
  }
}

@media only screen and (min-width:120rem) {
  .c0 {
    padding-left: 1.5rem;
    padding-right: 1.5rem;
  }
}

@media only screen and (min-width:1rem) {
  .c0 {
    max-width: 100%;
  }
}

@media only screen and (min-width:48rem) {
  .c0 {
    max-width: 100%;
  }
}

@media only screen and (min-width:64rem) {
  .c0 {
    max-width: 100%;
  }
}

@media only screen and (min-width:90rem) {
  .c0 {
    max-width: 90rem;
  }
}

@media only screen and (min-width:120rem) {
  .c0 {
    max-width: 90rem;
  }
}

<div
  className="c0"
/>
`;
